crates/kubelet/src/container/state.rs

```diff
@@ -0,0 +1,106 @@
+//! Functions for running Container state machines.
+use crate::container::{patch_container_status, Status};
+use crate::container::{Container, ContainerKey};
+use crate::pod::Pod;
+use crate::state::{ResourceState, SharedState, State, Transition};
+use chrono::Utc;
+use k8s_openapi::api::core::v1::Pod as KubePod;
+use kube::api::Api;
+use log::{debug, error, warn};
+
+/// Prelude for Pod state machines.
+pub mod prelude {
+    pub use crate::container::{Container, Handle, Status};
+    pub use crate::state::{ResourceState, SharedState, State, Transition, TransitionTo};
+}
+
+/// Iteratively evaluate state machine until it returns Complete.
+pub async fn run_to_completion<S: ResourceState<Manifest = Container, Status = Status>>(
+    client: &kube::Client,
+    initial_state: impl State<S>,
+    shared_state: SharedState<S::SharedState>,
```

Comment:
It is definitely confusing. I wonder if the `State` trait should actually be like `Handler`, and we might be able to just called it `shared` or something. Any ideas?

Reply:
Because this is billed as a "state machine" (which it is), I think the name "state" should remain with the actual state type. However, just using `shared` might be just fine

Reply:
Will do.